### ISSUE TITLE
╱ unicode pattern does not display as well as in the terminal

### ISSUE THREAD

**Issue by emmanueltouzery** [neovide/neovide#1197]
**Describe the bug**
Telescope can use the the unicode ╱ character (note, this is not the ascii /) to achieve an effect that diagonal lines are drawn on the screen. For instance for me, opening telescope to pick a file in the neovide project, and having it preview `assets/neovide.ico`, causes telescope to use the effect, while displaying the text "Binary cannot be previewed".
The effect works for me on the terminal, doesn't work in neovide though, where the symbols don't line up, at least partly due to vertical space between them.

I also get the issue when using diffs, because I configured this in nvim: `set fillchars+=diff:╱`. With that setup, deleted lines in diff also make use of that effect.

**To Reproduce**
1. install telescope
2. open neovide, open the neovide source
2. launch `:Telescope find_files`
3. preview `assets/neovide.ico`
4. compare display between terminal and neovide

**Expected behavior**
Same rendering in neovide and in neovim terminal.

**Screenshots**
![image](https://user-images.githubusercontent.com/339433/152640642-da8e9975-c53d-4bdd-a042-f5638decb309.png)
![image](https://user-images.githubusercontent.com/339433/152640681-7cb6c578-032a-4828-8b4f-d0061f718fed.png)

**Desktop (please complete the following information):**
 - OS: linux, fedora 35
 - Neovide 92bc1725f1733547eb0ae25b740425f03f358c2a
 - I'm using nerdfonts JetBrains Mono Nerd Fonts 2.2.0-rc from their github page

**Please run `neovide --log` and paste the contents of the `.log` file here:**
output was empty

**Comment by fredizzimo** [comment]
I think this could actually be related to https://github.com/neovide/neovide/issues/1198

If you look carefully at the two images, you can see that the Neovide floating Window is higher than the terminal one. In this case Neovide seems to have compensated by adding extra line spacing, but only in the floating Window. The line spacing in the normal buffer behind seems to be the same in both screenshots.

But just to verify, could you try to manually type that pattern in a normal buffer, and see if you get the extra line spacing there too?

**Comment by emmanueltouzery** [comment]
no, actually, i don't think it's related to floating windows. As I mentioned I can also reproduce this with diffs, because I set the diff fillchars.

Here is another case, no popup:

terminal:
![image](https://user-images.githubusercontent.com/339433/152643196-7346c884-4a1c-4826-bf62-2269b0ccca7e.png)

neovide:
![image](https://user-images.githubusercontent.com/339433/152643216-46fc78d4-7e12-4c8f-a470-be3fb57f6bf7.png)

**Comment by fredizzimo** [comment]
Ah, yes, I see, and I do have the same issue.

But I also tried the Windows Terminal, Windows Notepad and Alacritty and all of them renders the line spacing. So I'm wondering if the terminal emulator you are using is doing something special, which one are you using?

**Comment by emmanueltouzery** [comment]
hmm I'm using the Kitty terminal. But I checked with alacritty, I get a very slight gap. and with gnome-terminal it's seemless, same as with kitty.

alacritty:
![image](https://user-images.githubusercontent.com/339433/152644097-4e769684-147a-4b7a-bc55-d35a22448dbd.png)

**Comment by emmanueltouzery** [comment]
by the way, I checked, in telescope this parameter is set through `msg_bg_fillchar`, and this special "/" is the default:

```
          - msg_bg_fillchar:  Character to fill background of unpreviewable buffers with
                              Default: "╱"
```

**Comment by fredizzimo** [comment]
Are you sure that you use the same font in the terminal? I tried with `FiraCode NF` using that char, and I get this in Neovide
![image](https://user-images.githubusercontent.com/9946255/152644426-6d4dd051-9d9e-48bb-9a11-23a7d98344ba.png)

It's not perfect, there are some small antialiasing artifacts between the lines, but with JetBrainsMono NF, I get this
![image](https://user-images.githubusercontent.com/9946255/152644461-9d6cf17b-60f8-479d-97bc-485f3e2a0d79.png)

But for some reason Alacritty renders this with Jetbrains, not even the horizontal alignment is right 
![image](https://user-images.githubusercontent.com/9946255/152644567-5370bace-5f37-4a9f-bce9-229613a77bcb.png)

**Comment by emmanueltouzery** [comment]
I am using jetbrains mono in terminal and neovide. But I have a guess now maybe -- and keep in mind I know nothing about these things. What about if jetbrains mono does not offer that character.. so neovide falls back to plain /.. while the terminal renders that one character with another font, which actually knows that character?

**Comment by emmanueltouzery** [comment]
So, it could be related to this kind of issue, but I don't see what could be done based on this:
https://github.com/neovide/neovide/issues/75

**Comment by emmanueltouzery** [comment]
I ran `kitty --debug-font-fallback`. So it prints all the font fallbacks it does when rendering the screen. If there is 1 character on the screen for which a fallback is done, it will print one line. Not that character may not be the one you just typed! It's any character on screen.

Anyway I confirmed that kitty uses font fallback to Dejan Sans Mono to print a special character I have in my shell prompt. However it DOES NOT seem to use font fallback to display ╱

**Comment by fredizzimo** [comment]
JetBrainsMono should have it, or at least mine version does (the left is FiraCode and the right JetBrains Mono)
![image](https://user-images.githubusercontent.com/9946255/152644893-4d94816a-8426-4dbd-9d11-0f35f2d1ee67.png)

But if you look carefully, the version I have does not extend fully to the bottom, so I think there should be a gap. But I'm on Windows and I have built the font itself based on the latest available JetBrains version 2.242, and the latest Nerd font repository as of a maybe 3 weeks ago. It's possible that the RC version of Nerdfonts is slightly diffrerent though. 

If you want to check, you can install FontForge, and then `View/Goto` to go to the right glyph and open it.

**Comment by fredizzimo** [comment]
One interesting thing in the Jetbrains Mono version I have is that g extends further down than ╱
![image](https://user-images.githubusercontent.com/9946255/152645540-cd3c540b-0a50-4dbb-8300-31a05c9c0050.png)

**Comment by emmanueltouzery** [comment]
I see the same thing in fontforge than you. At least it seems like I may be able to edit the font manually to get the effect I want. I don't know which is the correct behavior between all these applications :(

**Comment by fredizzimo** [comment]
I'm not sure either. But it seems like the Jetbrains Mono should render without gaps, since it sticks up on the top part. But I think Neovide does not take that part of the font into account at all, and therefore renders the gaps.

So I think that Neovide should render a bigger box. But that does not in itself fix all issues, since if everything was rendered, the ends would still be antialiased, and look like there would be tiny gaps.

**Comment by emmanueltouzery** [comment]
For what it's worth, I have edited the nerdfonts 2.2RC version of jetbrains in fontforge and successfully got uninterrupted lines. But it's probably still worth checking whether neovide renders this correctly, I guess.

**Comment by cpmsmith** [comment]
> Anyway I confirmed that kitty uses font fallback to Deja Sans Mono to print a special character I have in my shell prompt. However it DOES NOT seem to use font fallback to display ╱

I believe the reason kitty is rendering this character differently, and not using a font fallback, is that it isn't using any font at all for that character. Like a lot of terminals, it has custom rendering of box drawing characters. You can see the definition for that particular character in the kitty source [here](https://github.com/kovidgoyal/kitty/blob/a5a7db6ce57078d7f6983880f542f083734f67c3/kitty/fonts/box_drawing.py#L845)

> with gnome-terminal it's seemless, same as with kitty.

I believe gnome-terminal does the same thing.

**Comment by fredizzimo** [comment]
Also, see this
* https://github.com/neovide/neovide/issues/1774

There has to be something strange with Jetbrains, I haven't seen these issues with other fonts

**Comment by fredizzimo** [comment]
There's no bug, it's just that the font width and/or height don't match exact pixels. It's possible that we calculated the ascender/descender and baseline wrong previously when I investigated, but that's not the case anymore.

We need custom box character drawing to fix the issue, see:
* https://github.com/neovide/neovide/issues/2491
